### ISSUE TITLE
Resque workers getting stuck at random places in the Job while enabling Elastic APM ruby agent.

### ISSUE THREAD

**Issue by bv-ritesh** [elastic/apm-agent-ruby#970]
## Describe the bug
We deployed Elastic APM ruby agent in our production servers. We have a very throughput cluster of servers(around 10) that are sending data to central APM server. After deploying we noticed that resque workers are getting stuck at random places in the jobs. 

We tried to reproduce it in a local/dev environment but no luck. The issue is persistent in the production environment.

## Environment

- **OS**: Ubuntu 18.04
- **Ruby version**: 2.3.3
- **Framework and version**: 5.2.2.1
- **APM Server version**: 7.10.2
- **Agent version**: 3.13.0

## Additional context

- Agent config options 
  <details>
    <summary>Click to expand</summary>

    ```
    replace this line with your agent config options
    remember to mask any sensitive fields like tokens
    ```
  </details>

**Comment by estolfo** [comment]
Hi @bv-ritesh, I'm sorry to hear you're having trouble using the apm ruby agent with Resque. 
Can you tell me more about your Resque jobs? What kind of tasks are they performing? Are they making other requests (to databases, other services, etc)? Have you seen any errors in the apm agent logs?
To answer your question, yes, you should set the `RUN_AT_EXIT_HOOKS` to `true`. As noted in our documentation [here](https://www.elastic.co/guide/en/apm/agent/ruby/current/supported-technologies.html#supported-technologies-resque), you must set this so that the agent is able to send all its events to the APM server when a fork exits.

**Comment by bv-ritesh** [comment]
Hi @estolfo , thank you for getting back to me. 
Our resque jobs are basically making requests to other websites, MySQL server, Redis Servers, etc. I have seen a few errors in APM agent logs like:

```
ElasticAPM] APM Server responded with an error:\n"{\"accepted\":2,\"errors\":[{\"message\":\"gzip: invalid header\"}]}\n"
[ElasticAPM] [PID:9716]: Queue is full (256 items), skipping…
[ElasticAPM] APM Server responded with an error:\n"{\"accepted\":0,\"errors\":[{\"message\":\"queue is full\"}]}\n"

```

Also I was not currently using RUN_AT_EXIT_HOOKS in our resque workers. Can this be causing the issue?

**Comment by estolfo** [comment]
Hi @bv-ritesh 
It's hard to say what effect not setting `RUN_AT_EXIT_HOOKS` would have had on your app exactly but I would suggest setting it and seeing if the behavior changes.
The error messages indicate that neither the agent nor the APM server are able to keep up with the volume of data collected in your app. I would suggest reading our documentation [here](https://www.elastic.co/guide/en/apm/agent/ruby/current/debugging.html#debugging-log-messages) for the error message `Queue is full (256 items), skipping…`.
I'd also suggest reading the APM server documentation [here](https://www.elastic.co/guide/en/apm/server/current/common-problems.html#queue-full) for the error message `APM Server responded with an error:\n"{\"accepted\":0,\"errors\":[{\"message\":\"queue is full\"}]}\n"`

What kind of tasks are your Resque workers performing? It sounds like there is a high volume of activity. I see you have a long list of disabled instrumentations in your config so perhaps there's another one that should be added to the list. Depending on what tasks you're performing and what's important for you to instrument, you also might want to adjust the `transaction_sample_rate`.

**Comment by bv-ritesh** [comment]
@estolfo thank you for your reply, I agree with you that we did not tweak in scale related settings for APM server and agent. But the question is that we want APM agent to work without these dependencies. APM agent should not affect our server functioning.
Is it possible to identify why our workers getting blocked at random places? that would help.

**Comment by estolfo** [comment]
@bv-ritesh Yes, I understand you'd like to have the APM agent work with no extra configuration. How do you observe the workers getting blocked? What do you see in the Resque logs?

**Comment by bv-ritesh** [comment]
@estolfo also adding more information, we are using one more gem with resque i.e  [resque-multi-job-forks](https://github.com/stulentsev/resque-multi-job-forks) which helps in controlling how many jobs per fork we want to execute.

  Yes we saw in resque logs that worker **Got** the Job but **done** is not coming for long time. also with our logs we can see that Job is not getting completed. we placed multiple logs in the Job to identify where Jobs are getting stuck but there is no pattern to it. It is getting blocked at random places.

Please let me know in case you need more information to debug this or help us.

**Comment by estolfo** [comment]
Have you tried not using `resque-multi-job-forks` and seeing if your app works better with the Elastic APM agent? I'm not familiar with the gem but I've looked at their documentation and I don't see a reason why there might be an issue.

I also don't want to ignore this error in one of your first comments: `gzip: invalid header`. Can you try setting the `http_compression` config to `false` and see if that changes anything?

And if you could provide any logs from Resque, that would be helpful. I'd like to see when the agent is being started, when the process is forked, and how it far along the jobs get.

**Comment by bv-ritesh** [comment]
@estolfo, running without resque-multi-job-forks is not possible in production and in development we are not facing this issue.
we will try with `http_compression: false` once
Also ill see if I can share the resque worker logs in future when we try it again.

**Comment by estolfo** [comment]
Hi @bv-ritesh Is this still an issue?

**Comment by bv-ritesh** [comment]
Hi @estolfo we didn't get further time to investigate this as this was happening on our production servers. Could it be linked to no of allowed threads on the server?

**Comment by estolfo** [comment]
Hi @bv-ritesh, can you explain what you mean by no allowed threads on the server?

**Comment by bv-ritesh** [comment]
Hi @estolfo allowed threads means maximum threads that can be created on a linux server.
